### PR TITLE
Devel

### PR DESCRIPTION
**Pull request by SamKG**
Compiled learning and forward computation for autodiff

### CODE REVIEW

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

**Comment by buildbot-princeton** [comment]
Can one of the admins verify this patch?

File: tests/composition/test_autodiffcomposition.py [hunk 1]
```diff
@@ -839,7 +908,14 @@ def test_and_training_time(self, eps, opt):
...
         ]
     )
-    def test_xor_training_time(self, eps, opt):
+    @pytest.mark.parametrize("mode", ['Python',
+                                    #pytest.param('LLVM', marks=pytest.mark.llvm),
+                                    #pytest.param('LLVMExec', marks=pytest.mark.llvm),
+                                    pytest.param('LLVMRun', marks=pytest.mark.llvm),
+                                    #pytest.param('PTXExec', marks=[pytest.mark.llvm, pytest.mark.cuda]),
+                                    #pytest.param('PTXRun', marks=[pytest.mark.llvm, pytest.mark.cuda])
+                                    ])
```
[the remark this answers is not in the document]
Comment (SamKG) [reply tests/composition/test_autodiffcomposition.py]: LLVMExec is not currently supported, as I require some extra parsing for the autodiff stimuli not present in comp_exec.execute

File: tests/composition/test_autodiffcomposition.py [hunk 2]
```diff
@@ -168,7 +176,14 @@ def test_training_then_processing(self):
...
     )
-    def test_various_loss_specs(self, loss):
+    @pytest.mark.parametrize("mode", ['Python',
+                                    #pytest.param('LLVM', marks=pytest.mark.llvm),
+                                    #pytest.param('LLVMExec', marks=pytest.mark.llvm),
+                                    pytest.param('LLVMRun', marks=[pytest.mark.llvm,pytest.mark.xfail]), # these loss specs remain unimplemented at the moment
+                                    #pytest.param('PTXExec', marks=[pytest.mark.llvm, pytest.mark.cuda]),
+                                    #pytest.param('PTXRun', marks=[pytest.mark.llvm, pytest.mark.cuda])
+                                    ])
+    def test_various_loss_specs(self, loss, mode):
```
[the remark this answers is not in the document]
Comment (SamKG) [reply tests/composition/test_autodiffcomposition.py]: It is used in the constructor for AutodiffComposition

**Comment by SamKG** [comment]
Has merge conflict with #1273. This one should be merged first - I have a working version with changes from #1273 merged in.

**Comment by jvesely** [comment]
retest this please

**Comment by jvesely** [comment]
retest this please

**Pull request merged**